raiden/raiden_service.py

```diff
@@ -719,18 +719,28 @@ def handle_state_changes(self, state_changes: List[StateChange]) -> List[Greenle
         # and try to DoS it, with the expectation that the victim would
         # forward the payment, but wouldn't be able to send a transaction to
         # the blockchain nor update a MS.
+
+        # Since several state_changes in one batch of state_changes can trigger
+        # the same PFSCapacityUpdate or MonitoringUpdate we want to iterate over
+        # all state changes to produce and send only unique messages. Assumption is
+        # that the latest related state_change defines the correct messages.
+        # Goal is to reduce messages.
+
+        batch_has_monitoring_update = False
```

Comment:
Here it would be better to have:

```python
latest_balance_proofs: Dict[ChannelID, BalanceProofStateChange] = dict()
```

And then after the `for state_change in state_changes:` you can iterate over the dictionary and send all known BPs.

Reply:
you mean like now?